src/kubernetes.cc

```diff
@@ -796,6 +796,15 @@ void KubernetesReader::WatchMaster(
   if (verbose) {
     LOG(INFO) << "WatchMaster(" << name << "): Contacting " << endpoint;
   }
+  std::mutex last_received_mutex;
+  auto last_received = std::chrono::high_resolution_clock::now();
+  auto timeout = std::chrono::seconds(config_.HealthCheckWatchTimeoutSeconds());
```

Comment:
Seems like you can have a single variable called `expiration` that is set to `std::chrono::high_resolution_clock::now() + std::chrono::seconds(config_.HealthCheckWatchTimeoutSeconds())` here and in the watcher callback, i.e.:
```c++
  std::mutex expiration_mutex;
  auto expiration = std::chrono::high_resolution_clock::now() +
                    time::seconds(config_.HealthCheckWatchTimeoutSeconds());
  ...
      [&expiration_mutex, &expiration]{
        std::lock_guard<std::mutex> expiration_lock(expiration_mutex);
        return std::chrono::high_resolution_clock::now() < expiration;
      }
  ...
      [=, &expiration_mutex, &expiration](json::value raw_watch) {
        {
          std::lock_guard<std::mutex> expiration_lock(expiration_mutex);
          expiration = std::chrono::high_resolution_clock::now() +
                       time::seconds(config_.HealthCheckWatchTimeoutSeconds());
        }
        WatchEventCallback(callback, name, std::move(raw_watch));
      }
```

Might be a good idea to factor out `config_.HealthCheckWatchTimeoutSeconds()` into a local variable called `timeout`, something like:
```c++
  const int timeout = config_.HealthCheckWatchTimeoutSeconds();
  auto expiration = std::chrono::high_resolution_clock::now() + time::seconds(timeout);
  ...
      [&expiration_mutex, &expiration]{
        std::lock_guard<std::mutex> expiration_lock(expiration_mutex);
        return std::chrono::high_resolution_clock::now() < expiration;
      }
  ...
      [=, &expiration_mutex, &expiration](json::value raw_watch) {
        {
          std::lock_guard<std::mutex> expiration_lock(expiration_mutex);
          expiration = std::chrono::high_resolution_clock::now() + time::seconds(timeout);
        }
        WatchEventCallback(callback, name, std::move(raw_watch));
      }
```

Reply:
Done, thanks.